src/common/fast-path.js

```diff
@@ -2,13 +2,23 @@
 
 const assert = require("assert");
 
+/**
+ * @class
+ * @constructor
+ * @template {string} T
+ * @property {T[]} stack
```

Comment:
Why make this class generic? Isn't `value` always a string?

Reply:
I used this as a start base:

https://github.com/ffflorian/DefinitelyTyped/blob/14f6f073a73f28fcdee4d899b50254eac30d4aac/types/prettier/index.d.ts#L13

Reply:
Oh, so `T` is an AST node, not a string. `@template {string} T` means `<T exends string>`.

Reply:
Yes

Reply:
So, this constraint is incorrect and also the type for `stack` is incorrect.

Reply:
You think the definition of `DefinitelyTyped` is wrong or `string` is wrong?
Can you be more precisely, what is wrong?

Reply:
Those definitions look fine. But if you are simply copying the types from there, the parts `@template {string} T` and `@property {T[]} stack` don't correspond to the source.

Reply:
So should it be `string` everywhere instead of `T`?

Reply:
You should either 1) read and understand the code and add types based on your analysis, or 2) at least copy the types from those definitions precisely without adding random bits. Sorry if i'm harsh, but adding types without understanding the code is a really questionable endeavor.

Reply:
2) But I cant just copy those definitions because there are not JSDoc compatible :(

So I tried using some of both worlds
But it seams that this causes problems, so I will rework this and only use `string` at this point

(Only need to be at home for this)

Reply:
"and only use string at this point" - that's what I called 'random bits', unfortunately.

If you have read the code, what type are the elements of the `stack` array?

Reply:
I think it should be something like `Array<AstNode>`
_Whatever AstNode is_

Reply:
How did you come to such a conclusion?

Reply:
Here is one example:

https://github.com/prettier/prettier/blob/master/src/common/util.js#L621-L628

Reply:
Here is a [counterexample](https://github.com/prettier/prettier/blob/master/src/common/fast-path.js#L10-L16). And there are comments in the code of the class that describe the structure of the stack.

Also you seem to have not read the [docs](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html) on JSDoc support in TS. Otherwise you wouldn't use the `@property` tag the way you did. It works only with [typedefs](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#typedef-callback-and-param). "Properties are inferred from assignments in class bodies" as [it is said in the very beginning](https://www.typescriptlang.org/docs/handbook/type-checking-javascript-files.html#properties-are-inferred-from-assignments-in-class-bodies) of that doc page. To override the inference, you can use `@type`. Also there is no need to add both `@class` and `@constructor`, they are synonyms... I could go on, but I don't really have time for that.

I've got mixed feelings about this PR. On the one hand, it might become a good enough starting point for a subsequent gradual migration to TS. On the other, a migration like that requires a deep understanding of both TS and the code base. Sloppy types will quickly become a nuisance like it happened with Redux, whose maintainers now [regret](https://twitter.com/acemarke/status/1149795712206594048) they have messed with TS without having a good expertise in it.

Reply:
I don't know how old and experienced you are, but I'm only 26 years old and have 3 years of JavaScript experience (2 of these years in TypeScript)
All my experiences are based on es6+

So instead of telling me all the time that I'm doing it wrong, you could be nice and help me get it right
That would provide a much more enjoyable community experience

That said,
sure, I can set it to `string`, but I'm confused by the `DefinitelyTyped` definitions

And also yes, I know TypeScript and I'm not very into JSDoc, because I don't need it my self (for types) because I'm working with `.ts`-native files in all my projects. So I have approximately 10 days experience in JSDoc